Formatting.md

```diff
@@ -0,0 +1,158 @@
+- JavaScript uses backticks to denote template strings which support direct expression substitution. No special formatting syntax is provided.
+- C++ does funny things with [`<<`](http://www.cplusplus.com/reference/ostream/ostream/operator%3C%3C/). There are also various 3rd party libraries
+  * [fastformat](http://www.fastformat.org/)
+  * [fmtlib](https://github.com/fmtlib/fmt)
+- Rust [std::fmt](https://doc.rust-lang.org/std/fmt/) uses positional and/or keyword arguments, with its own minilanguage (appears similar to Python's).
+- Haskell [`printf`](https://hackage.haskell.org/package/base-4.10.1.0/docs/Text-Printf.html) uses positional substitution, and usual printf formatting
+- C# has 2 formats
+  * [`String.Format`](https://msdn.microsoft.com/en-us/library/system.string.format(v=vs.110).aspx) uses positional substitution
+  * [Interpolated strings](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/interpolated-strings) provide direct substitution, with its own minilanguage for formatting.
+- Common lisp [`Format`](https://en.wikipedia.org/wiki/Format_(Common_Lisp)) uses positional substitution, but its minilanguage provides support for things like iteration and recursion (and is quite possibly Turing complete).
+- Swift has
+  - [`String`](https://developer.apple.com/documentation/swift/string) supports positional substitution with `printf`-style formatting
+  - [direct substitution](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/StringsAndCharacters.html#//apple_ref/doc/uid/TP40014097-CH7-ID292), but no formatting (suggestions to get around this rely on [calling the former inside the latter](http://en.swifter.tips/output-format/)).
+
+## Proposal
+
+The general proposal is to specify the format via additional arguments to interpolated strings. This can utilise either keyword or positional arguments (or no arguments for default printing):
+
+    "π is approximately $pi."
+    "π is approximately $(pi, fracdigits=4)."
+    "π is approximately $(pi, Decimal(4))."
+
+The multiple-argument ones are currently syntax errors, so will not present backward compatibility problems. The intention is that keyword arguments expose the straightforward features, and more complicated features and extensions can be provided by Julia objects passed as second arguments.
+
+### Implementation and performance considerations
+
+Interpolation would call a function `stringformat`. The above would each call something like
+
+    string("π is approximately ", stringformat(pi), ".")
+    string("π is approximately ", stringformat(pi, fracdigits=4), ".")
+    string("π is approximately ", stringformat(pi, Decimal(4)), ".")    
+
+where `stringformat` is the appropriately defined function returning a formatted string.
+
+However this may result in large numbers of intermediate allocations of small strings. Ideally we would want this to lower to something like
```

Comment:
Interpolation could instead call `stringformat(io, pi, fracdigits=4)` on the `io::IOBuffer` that is used for string construction by `string`.

Maybe `"π is approximately $(pi, fracdigits=4)."` could instead lower `string("π is approximately ", StringFormat(pi, fracdigits=4), ".")`, where a `f::StringFormat{T,KW}` object caches both the object `pi::T` and a namedtuple `(fracdigits=4,)::KW`, and has a `print(io, f)` method (used by both `string` and `print`).

Reply:
i.e.
```jl
immutable StringFormat{T,KW}
    x::T # the value to be printed
    kws::KW # NamedTuple of keyword arguments
end
write(io::IO, f::StringFormat) = write(io, f.x) # fallback method
```

Reply:
That's a good idea.

Reply:
this wouldn't immediately help with the `print(io, "π is approximately $(pi, fracdigits=4).")` case though.

Reply:
You would pass `StringFormat(...)` in that case. 

I don’t think we’d want $ there

Reply:
Wait, I misread it. Yes, you’d write twice there. Not sure it’s a big deal.

Reply:
One option would be to lower to `LazyString(...)` instead of `string(...)`